src/main/java/de/tum/in/www1/artemis/service/programming/JavaTemplateUpgradeService.java

```diff
@@ -0,0 +1,217 @@
+package de.tum.in.www1.artemis.service.programming;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import org.apache.maven.model.Dependency;
+import org.apache.maven.model.Model;
+import org.apache.maven.model.Plugin;
+import org.apache.maven.model.io.xpp3.MavenXpp3Reader;
+import org.apache.maven.model.io.xpp3.MavenXpp3Writer;
+import org.codehaus.plexus.util.xml.pull.XmlPullParserException;
+import org.eclipse.jgit.api.errors.GitAPIException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.core.io.Resource;
+import org.springframework.stereotype.Service;
+
+import de.tum.in.www1.artemis.ResourceLoaderService;
+import de.tum.in.www1.artemis.domain.ProgrammingExercise;
+import de.tum.in.www1.artemis.domain.Repository;
+import de.tum.in.www1.artemis.domain.enumeration.RepositoryType;
+import de.tum.in.www1.artemis.service.ProgrammingExerciseService;
+import de.tum.in.www1.artemis.service.UserService;
+import de.tum.in.www1.artemis.service.connectors.GitService;
+
+/**
+ * Service for upgrading of Java template files
+ */
+@Service
+public class JavaTemplateUpgradeService implements TemplateUpgradeService {
+
+    private final Logger log = LoggerFactory.getLogger(JavaTemplateUpgradeService.class);
+
+    private final ProgrammingExerciseService programmingExerciseService;
+
+    private final GitService gitService;
+
+    private final UserService userService;
+
+    private final ResourceLoaderService resourceLoaderService;
+
+    public JavaTemplateUpgradeService(ProgrammingExerciseService programmingExerciseService, GitService gitService, ResourceLoaderService resourceLoaderService,
+            UserService userService) {
+        this.programmingExerciseService = programmingExerciseService;
+        this.gitService = gitService;
+        this.userService = userService;
+        this.resourceLoaderService = resourceLoaderService;
+    }
+
+    @Override
+    public void upgradeTemplate(ProgrammingExercise exercise) {
+        // TODO: Support sequential test runs
+        if (exercise.hasSequentialTestRuns()) {
+            return;
+        }
+        // Template and solution repository can also contain a project object model for some project types
+        updateRepository(exercise, RepositoryType.TEMPLATE.getName(), RepositoryType.TEMPLATE);
+        updateRepository(exercise, RepositoryType.SOLUTION.getName(), RepositoryType.SOLUTION);
+        updateRepository(exercise, "test/projectTemplate", RepositoryType.TESTS);
```

Comment:
why are we not using here `RepositoryType.TEST.getName()`? Could you add a comment?

Reply:
RepositoryType.Test.getName() -> "tests". So it doesn't correspond to the folder structure of the templates. The other reason is that tests/ contains two pom.xml. One is used for sequential tests only. As this isn't supported now I use the more specific path to the right pom. I'm already working on a follow up which restructures the code slightly, so I don't want to add a comment for this.